File: ml-benchmark/autolifter/single-pass/max_1s_p.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = 0;
Config SampleIntMax = 1;
Config SampleSize = 20;

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> {Int, Int}.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

max1s_with_pos = fix (\f: Int -> Int -> List -> {Int, Int}. \pre: Int. \i: Int. \xs: List.
  match xs with
    nil _ ->
      let len = - i pre in
        {len, pre}
  | cons {h, t} ->
      if == h 1 then f pre (+ i 1) t
      else let len = - i pre in
        let res = f (+ i 1) (+ i 1) t in
          if >= len res.1 then {len, pre} else res
  end
) 0 0;

main = single_pass max1s_with_pos;

File: ml-benchmark/autolifter/single-pass/max_1s_p.ml
type lst = Cons of int * lst | Nil

let single_pass v =
  let rec run xs =
    match xs with
    | Nil -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let inf = 100

let max a b =
  if a < b then b else a

let rec max1s_with_pos pre i xs =
  match xs with
  | Nil ->
      let len = i - pre in
      (len, pre)
  | Cons (h, t) ->
      if h = 1 then
        max1s_with_pos pre (i + 1) t
      else
        let len = i - pre in
        let res = max1s_with_pos (i + 1) (i + 1) t in
        match res with
        | (rlen, rpre) ->
            if len >= rlen then (len, pre) else (rlen, rpre)

let program = single_pass max1s_with_pos